aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -2805,17 +2805,36 @@ static void linalg_lstsq_out_info(
   // now the actual call that computes the result in-place (apply_lstsq)
   at::_lstsq_helper(solution, rank, singular_values, infos, input_working_copy, rcond, driver);
 
+  // residuals are available only if m > n and drivers other than gelsy used
   if (m > n && driver != "gelsy") {
-    // LAPACK stores residuals data for postprocessing in rows n:(m-n)
-    auto raw_residuals = solution.narrow(/*dim=*/-2, /*start=*/n, /*length*/m - n);
-    if (raw_residuals.is_complex()) {
-      // in-place abs is not supported for complex tensors
-      raw_residuals = raw_residuals.abs();
-      raw_residuals.pow_(2);
-    } else {
-      raw_residuals.abs_().pow_(2);
+    // if the driver is gelss or gelsd then the residuals are available only if rank == n
+    bool compute_residuals = true;
+    if (driver != "gels") {
+      if (input.dim() == 2) {
+        compute_residuals = (rank.item().toInt() == n);
+      } else {
+        // if any of the computed ranks in equal to n then do the computation of residuals
+        compute_residuals = at::any(rank == n).item().toBool();
+      }
+    }
+    if (compute_residuals) {
+      // LAPACK stores residuals data for postprocessing in rows n:(m-n)
+      auto raw_residuals = solution.narrow(/*dim=*/-2, /*start=*/n, /*length*/m - n);
+      if (raw_residuals.is_complex()) {
+        // in-place abs is not supported for complex tensors
+        raw_residuals = raw_residuals.abs();
+        raw_residuals.pow_(2);
+      } else {
+        raw_residuals.abs_().pow_(2);
+      }
+      at::sum_out(residuals, raw_residuals, /*dim=*/-2, /*keepdim=*/false, /*dtype*/real_dtype);
+
+      // if at::any(rank != n) then the residuals for i-th batch with rank[i] != n are not available
+      // and the computed result is not meaningful, we fill it with INFINITY as we cannot resize it to (0,) shape
+      if (driver != "gels") {
+        residuals.masked_fill_((rank != n).unsqueeze_(-1), INFINITY);
```

Comment:
When m < n and rank == m there are infinite number of solution satisfying b - Ax = 0. Probably residuals should be filled with zero here and not infinity.

Reply:
@Lezcano @nikitaved, would you like to weigh-in on this UX?

Reply:
Unique or not unique solution, in this case the residual is expected to be very close to zero.

Reply:
Residual is available as a byproduct of the LAPACK calls after some post-processing (abs and pow are fast element-wise operations). Raw residual is available if gels driver is used for any input and only for rank==n inputs if gelss, gelsd drivers are used. So for other cases calculating the residual would require matrix-matrix multiplication `ax` and then calculating the norm, I think it's an unnecessarily expensive procedure, and it's not immediately clear we should do that.

Reply:
I'd prefer NumPy and SciPy did not have this return at all 😔

Reply:
I may have not understood this, but this is under the `m > n` condition, right?. In that case, there isn't in general a solution, and you're always going to have some residuals. I agree with @nikitaved that it would be reasonable to return ``||ax-b||^2_F``. It is true that it performs a matrix-matrix multiplication, but I would assume that that'd be alright when compared with how expensive this operation already is (it performs an SVD).

On the other hand, for matrices that are not full-rank, this function is not differentiable, and since this is under ``if (driver != "gels")``, it means that this will be hit just by CPU users. If a user is using CPU and tries to differentiate this, it will break. As such, what we return here only matters for people that use CPU in a non-differentiable setting. In this setting a matrix-matrix multiplication is quite costly on its own...

I'd often prefer to lean towards correctness, but I completely agree that this is a return that's not particularly useful in general and it'd be best removed.

Reply:
Not returning residuals is what I actually proposed while working on `linalg.lstsq` :)